src/java/org/apache/cassandra/transport/Dispatcher.java

```diff
@@ -110,6 +114,7 @@ else if (backpressure == Overload.BYTES_IN_FLIGHT)
         Message.logger.trace("Received: {}, v={}", request, connection.getVersion());
         connection.requests.inc();
         Message.Response response = request.execute(qstate, queryStartNanoTime);
+        CoordinatorTrackWarnings.done();
```

Comment:
Is `done()` not called if there is an exception thrown from `execute()`. One thing we might consider is looking at moving the `init()`/`done()` pair to `Message.Request#execute()`, where tracing is enabled. (i.e. There is already some plumbing there to do something conceptually similar for tracing, and tbh, if we ever try to have a more unified thread-local query context, that would make things easier to pull together.)

Reply:
> Is done() not called if there is an exception thrown from execute()

ATM no, nor do we retry ClientWarn in those cases as thats handled 2 lines below.  If you look at `org.apache.cassandra.transport.Message.Request#execute(org.apache.cassandra.service.QueryState, long)` its marked `final` and does imply exceptions could go up (it does try-finally; no catch).  If I look at the caller of this dispatch logic it actually handles the exception and clears ClientWarn; will do the same here.

Reply:
pushing a patch which handles this higher up, avoids double publishing, and will always clear the map